metals/src/main/scala/scala/meta/internal/metals/CodeLensProvider.scala

```diff
@@ -1,50 +1,114 @@
+import scala.meta.internal.metals.CodeLensProvider._
 import scala.meta.internal.metals.MetalsEnrichments._
 import scala.meta.internal.mtags.Semanticdbs
 import scala.meta.io.AbsolutePath
 
 final class CodeLensProvider(
-    classes: BuildTargetClasses,
+    buildTargetClasses: BuildTargetClasses,
     buffers: Buffers,
     buildTargets: BuildTargets,
+    compilations: Compilations,
     semanticdbs: Semanticdbs
-) {
-  def findLenses(path: AbsolutePath): util.List[l.CodeLens] = {
-    buildTargets.inverseSources(path) match {
-      case Some(buildTarget) if classes.main(buildTarget).nonEmpty =>
-        findLenses(path, buildTarget).asJava
-      case _ =>
-        emptyList[l.CodeLens]()
-    }
+)(implicit ec: ExecutionContext) {
+  // code lenses will be refreshed after compilation or when workspace gets indexed
+  def findLenses(path: AbsolutePath): Seq[l.CodeLens] = {
+    val lenses = buildTargets
+      .inverseSources(path)
+      .filterNot(compilations.isCurrentlyCompiling)
+      .map { buildTarget =>
+        val classes = buildTargetClasses.classesOf(buildTarget)
+        val lenses = findLenses(path, buildTarget, classes)
+        lenses
+      }
+
+    lenses.getOrElse(Nil)
   }
 
   private def findLenses(
       path: AbsolutePath,
-      buildTarget: b.BuildTargetIdentifier
-  ): List[l.CodeLens] = {
+      target: b.BuildTargetIdentifier,
+      classes: BuildTargetClasses.Classes
+  ): Seq[l.CodeLens] = {
     semanticdbs.textDocument(path).documentIncludingStale match {
       case Some(textDocument) =>
         val distance =
           TokenEditDistance.fromBuffer(path, textDocument.text, buffers)
-        val mainClasses = classes.main(buildTarget)
 
-        val lenses = for {
+        for {
           occurrence <- textDocument.occurrences
-          if mainClasses.contains(occurrence.symbol)
-          mainClass = mainClasses(occurrence.symbol)
+          if occurrence.role.isDefinition
+          symbol = occurrence.symbol
+          commands = {
+            val main = classes.mainClasses
+              .get(symbol)
+              .map(MainClassLensFactory.commands(target, _))
+              .getOrElse(Nil)
+            val tests = classes.testSuites
+              .get(symbol)
+              .map(TestSuitesLensFactory.commands(target, _))
+              .getOrElse(Nil)
+            main ++ tests
+          }
+          if commands.nonEmpty
           range <- occurrence.range
             .flatMap(r => distance.toRevised(r.toLSP))
             .toList
-          arguments = List(buildTarget.getUri, mainClass.getClassName)
-        } yield
-          new l.CodeLens(range, ClientCommands.RunMain.toLSP(arguments), null)
-        lenses.toList
+          command <- commands
+        } yield new l.CodeLens(range, command, null)
       case _ =>
         Nil
     }
   }
 }
+
+object CodeLensProvider {
+  import JsonParser._
+  sealed trait CommandFactory[A] {
```

Comment:
What do you think about implementing this class like this instead?

```suggestion
case class CommandFactory[A](names: List[String], kind: String, arg: A => AnyRef) {
  def commands(..) = ???
}
val mainCommandFactory = CommandFactory[b.ScalaMainClass](List("run"), b.DebugSessionParamsDataKind.SCALA_MAIN_CLASS, arg => arg)
val testCommandFactory = CommandFactory[String](List("run"), b.DebugSessionParamsDataKind.SCALA_TEST_CLASS, arg => List(arg).asJava)
```

I personally would find such an  implementation easier to read and extend compared to the current trait design.